### PR TITLE
clib.converison._to_numpy: Add tests for pandas.Series with pandas numeric dtypes

### PR DESCRIPTION
**Pull request by seisman**
This PR adds tests for pandas.Series with pandas/pyarrow numeric dtypes (in 7222db23bbdfbb298b211c02ab981e71ab433be6).

Tests pass with pandas 2.2 but fail with pandas 2.1. Check https://github.com/GenericMappingTools/pygmt/actions/runs/11714111582/job/32628172796?pr=3584 for details.

The failures are due to upstream pandas v2.2 changes. Previously, all pandas nullable dtypes are converted to np.object_ dtype. Since pandas v2.2, the new rules are (source: https://pandas.pydata.org/docs/whatsnew/v2.2.0.html#to-numpy-for-numpy-nullable-and-arrow-types-converts-to-suitable-numpy-dtype):

> - float dtypes are cast to NumPy floats
> - integer dtypes without missing values are cast to NumPy integer dtypes
> - integer dtypes with missing values are cast to NumPy float dtypes and NaN is used as missing value indicator
> - boolean dtypes without missing values are cast to NumPy bool dtype
> - boolean dtypes with missing values keep object dtype
> - datetime and timedelta types are cast to Numpy datetime64 and timedelta64 types respectively and NaT is used as missing value indicator

Following the first three points, we add the workaround for pandas<2.2 to explicitly specify how to map pandas dtypes into numpy dtypes (a4f15cd778b146cc787450ba81b36da81a3a2efd). It is an improved version of PR #3505.

Here is the minimal example to understand the behavior change:

**With pandas 2.1**
```python
In [1]: import numpy as np

In [2]: import pandas as pd

In [3]: x = pd.Series([1, 2, 3], dtype=pd.Int32Dtype())

In [4]: x.dtype
Out[4]: Int32Dtype()

In [5]: np.ascontiguousarray(x)
Out[5]: array([1, 2, 3], dtype=object)

In [6]: x = pd.Series([1, pd.NA, 3], dtype=pd.Int32Dtype())

In [7]: np.ascontiguousarray(x)
Out[7]: array([1, <NA>, 3], dtype=object)
```
**With pandas 2.2**
```python
In [1]: import numpy as np

In [2]: import pandas as pd

In [3]: x = pd.Series([1, 2, 3], dtype=pd.Int32Dtype())

In [4]: np.ascontiguousarray(x)
Out[4]: array([1, 2, 3], dtype=int32)

In [5]: x = pd.Series([1, pd.NA, 3], dtype=pd.Int32Dtype())

In [6]: np.ascontiguousarray(x)
Out[6]: array([ 1., nan,  3.])
```

Related to #3581, #2848, #3513, #3583.

~~Wait for #3583.~~

### CODE REVIEW

**Comment by seisman** [comment]
As explained in the top post, currently this PR does two things:

- Add tests for pandas dtypes
- Improve the pandas->numpy dtype conversion for pd.Series with NA values (pandas<2.2)

After finishing this PR, we have two options:

1. Change the PR title to something like "Improve the conversion of pandas dtypes with missing values for pandas<=2.1" and mark it as a "bug".
2. Cherry-pick the changes in a4f15cd778b146cc787450ba81b36da81a3a2efd, and open a separate PR so that we'll have two commits in the main branch, one for the new workaround, and one for the newly added tests.

I'm inclined to option 2.

**Comment by weiji14** [comment]
> 2\. Cherry-pick the changes in [a4f15cd](https://github.com/GenericMappingTools/pygmt/commit/a4f15cd778b146cc787450ba81b36da81a3a2efd), and open a separate PR so that we'll have two commits in the main branch, one for the new workaround, and one for the newly added tests.
> 
> 
> I'm inclined to option 2.

Yeah, agree to option 2. Keep the unit tests here, and open a separate PR with the code changes from a4f15cd778b146cc787450ba81b36da81a3a2efd

File: pygmt/clib/conversion.py
Comment (seisman) [inline pygmt/clib/conversion.py:191]: For pandas objects with missing values, we can't just map `Int32` to `np.float64`, as shown below (with pandas 2.1):
```python
In [1]: import pandas as pd

In [2]: import numpy as np

In [3]: x = pd.Series([1, None, 3], dtype=pd.Int32Dtype())

In [4]: np.ascontiguousarray(x, dtype=np.float64)
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
Cell In[4], line 1
----> 1 np.ascontiguousarray(x, dtype=np.float64)

File ~/opt/miniforge/envs/pygmt/lib/python3.12/site-packages/pandas/core/series.py:953, in Series.__array__(self, dtype)
    906 """
    907 Return the values as a NumPy array.
    908
   (...)
    950       dtype='datetime64[ns]')
    951 """
    952 values = self._values
--> 953 arr = np.asarray(values, dtype=dtype)
    954 if using_copy_on_write() and astype_is_view(values.dtype, arr.dtype):
    955     arr = arr.view()

File ~/opt/miniforge/envs/pygmt/lib/python3.12/site-packages/pandas/core/arrays/masked.py:575, in BaseMaskedArray.__array__(self, dtype)
    570 def __array__(self, dtype: NpDtype | None = None) -> np.ndarray:
    571     """
    572     the array interface, return my values
    573     We return an object array here to preserve our scalar values
    574     """
--> 575     return self.to_numpy(dtype=dtype)

File ~/opt/miniforge/envs/pygmt/lib/python3.12/site-packages/pandas/core/arrays/masked.py:487, in BaseMaskedArray.to_numpy(self, dtype, copy, na_value)
    481 if self._hasna:
    482     if (
    483         dtype != object
    484         and not is_string_dtype(dtype)
    485         and na_value is libmissing.NA
    486     ):
--> 487         raise ValueError(
    488             f"cannot convert to '{dtype}'-dtype NumPy array "
    489             "with missing values. Specify an appropriate 'na_value' "
    490             "for this dtype."
    491         )
    492     # don't pass copy to astype -> always need a copy since we are mutating
    493     with warnings.catch_warnings():

ValueError: cannot convert to 'float64'-dtype NumPy array with missing values. Specify an appropriate 'na_value' for this dtype.
```
So we have to call `.to_numpy()` explicitly.

File: pygmt/clib/conversion.py
Comment (seisman) [inline pygmt/clib/conversion.py:190]: The expected dtype is `np.float64` except for `Float32`.

**Comment by seisman** [comment]
> > 2. Cherry-pick the changes in [a4f15cd](https://github.com/GenericMappingTools/pygmt/commit/a4f15cd778b146cc787450ba81b36da81a3a2efd), and open a separate PR so that we'll have two commits in the main branch, one for the new workaround, and one for the newly added tests.
> > I'm inclined to option 2.
> 
> Yeah, agree to option 2. Keep the unit tests here, and open a separate PR with the code changes from [a4f15cd](https://github.com/GenericMappingTools/pygmt/commit/a4f15cd778b146cc787450ba81b36da81a3a2efd)

I've opened PR #3596 with changes from a4f15cd, but please review changes in this PR. Any changes in `pygmt/clib/conversion.py` will be cherry-picked into PR #3596 after this PR is approved.

File: pygmt/clib/conversion.py
Comment (seisman) [inline pygmt/clib/conversion.py:191]: When dealing with missing values with pandas 2.1, we have some choices to convert pandas array into numpy array.

1. `np.ascontinuous(data, dtype=np.float64)`
2. `data.to_numpy(dtype=dtype, na_value=np.nan)`
3. `data.to_numpy(na_value=np.nan, dtype=dtype)`

As shown below, it turns out only the 3rd way works for all cases.

```python
In [1]: import pandas as pd
x =
In [2]: x = pd.Series([1, 2, pd.NA], dtype=pd.Int32Dtype())

In [3]: import numpy as np

In [4]: np.ascontiguousarray(x, dtype=np.float64)
...
ValueError: cannot convert to 'float64'-dtype NumPy array with missing values. Specify an appropriate 'na_value' for this dtype.
In [5]: x.to_numpy(dtype=np.float64, na_value=np.nan)
Out[5]: array([ 1.,  2., nan])

In [6]: x.to_numpy(na_value=np.nan).astype(np.float64)
Out[6]: array([ 1.,  2., nan])

In [7]: x = pd.Series([1, 2, pd.NA], dtype="float16[pyarrow]")

In [8]: x
Out[8]:
0     1.0
1     2.0
2    <NA>
dtype: halffloat[pyarrow]

In [9]: x.to_numpy(dtype=np.float64, na_value=np.nan)
...
ArrowTypeError: Expected np.float16 instance

In [10]: x.to_numpy(na_value=np.nan).astype(np.float64)
Out[10]: array([ 1.,  2., nan])
```

**Comment by seisman** [comment]
Commit 0615b5b7c2ec2a9716ef5d64c0dbf0f44f3fd37d catches bugs of the improved workaround for pandas<2.2. Will see what's happening and how to make the workaround more robust.

**Comment by seisman** [comment]
I prefer to have other PRs like #3610, #3609 reviewed and merged first, to ensure the workaround in this PR is robust and has no side-effects on other dtypes.

**Review by michaelgrund — APPROVED** [review]

**Comment by seisman** [comment]
@weiji14 I'm wondering if you have time to give this PR a final review.

**Review by weiji14 — APPROVED** [review]

**Pull request merged**
